packages/playwright-test/src/loader.ts

```diff
@@ -152,12 +152,28 @@ export class Loader {
     }
     this._fullConfig.metadata = takeFirst(config.metadata, baseFullConfig.metadata);
     this._fullConfig.projects = (config.projects || [config]).map(p => this._resolveProject(config, this._fullConfig, p, throwawayArtifactsPath));
+    this._assignUniqueProjectIds(this._fullConfig.projects);
+  }
+
+  private _assignUniqueProjectIds(projects: FullProjectInternal[]) {
+    const usedNames = new Set();
+    for (const p of projects) {
+      const name = p.name || 'project';
+      for (let i = 0; i < projects.length; ++i) {
+        const candidate = name + (i ? '#' + i : '');
+        if (usedNames.has(candidate))
+          continue;
+        p._id = candidate;
+        usedNames.add(candidate);
+        break;
+      }
+    }
   }
```

Comment:
I don't love how sensitive these IDs are. If the user wants to rename a project (or a test…or nest in a describe, etc.)—maybe fix a typo—all the IDs change, so the notion these are historical IDs are pretty fragile.

I think this is fine for now, but I suggest being more explicit in the docs. If requested, we could always add an ID generation function that can be customized (or allow more explicitly setting IDs for projects, tests, etc.).

Reply:
Yeah, we can always add a project.historyID!